README.md
# Linux Stack Made Easy.

The LEMP / LAMP software stack is a group of software that can be used
to serve dynamic web pages and web applications. This is an acronym
that describes a Linux operating system, with an Nginx or Apache web
server. The backend data is stored in the MySQL/MariaDB database and
or PostgreSQL as optional, and the dynamic processing is handled by
PHP, Python, Nodejs, etc.

## Prerequisites

- Fresh installation of Ubuntu 16.04 or 18.04 LTS.
- Domain with already pointed IP address to your server.

If you prefer to use Debian, you can check [`debian`](//github.com/riipandi/stackup/tree/debian) branch.

## Quick Start

Run this command as root and follow the wizard:

```sh
# Stable channel
bash <(curl -sLo- https://git.io/fhiA7 || wget -qO- https://git.io/fhiA7)

# Master branch
bash <(curl -sLo- https://git.io/fhiA7 || wget -qO- https://git.io/fhiA7) --dev
```

If you prefer to run installation manually, you just clone this repo then run provided scripts in `installer` directory.

Installation information stored at: `/usr/local/share/stackup.info`

### Installation notes in AWS

AWS Lightsail doesn't use password by default for ssh authentication.
Download SSH key from Lightsail management console, then:

```bash
cat LightsailDefaultKey-zone.pem > ~/.ssh/id_rsa_aws_oregon

ssh given_username@ip_address -i ~/.ssh/id_rsa_aws_oregon
```

Also, AWS Lightsail use generated hostname for you instance. So, you will 
need change your instance hostname:

```bash
hostnamectl set-hostname myhostname.mydomain.tld
```

If you create new sudo user, you will need execute this command:

```bash
sudo cp /home/ubuntu/.ssh/authorized_keys /home/new_username/.ssh/authorized_keys
sudo chmod 0600 /home/new_username/.ssh/authorized_keys
sudo chown new_username /home/new_username/.ssh/authorized_keys
```

Or, you can use your own SSH key.

## Some Useful Snippets

| Command             | Description                                         | Example Usage
| :------------------ | :-------------------------------------------------- | :------------
| create-buddy        | Create new user for application deployment          | `sudo create-buddy`
| create-user         | Create new user with sudo privileges                | `sudo create-user`
| set-default-php     | Set default PHP version                             | `sudo set-php 7.2`
| set-default-python  | Set default Python version                          | `sudo set-python 3.5`
| site-ghost          | Create or uninstall Ghost blogging platform         | `site-ghost example.com`
| ------------------- | --------------------------------------------------- | `site-ghost example.com --uninstall`
| site-php            | Create Nginx virtualhost for php-fpm backend        | `sudo site-php example.com`
| site-proxy          | Create Nginx virtualhost for reverse proxy          | `sudo site-proxy example.com`
| site-python         | Create Nginx virtualhost for python backend         | `sudo site-python example.com`
| fix-permission      | Fix directory and file permission                   | `sudo fix-permission /my/path user:group`
| mysql-create        | Create new MySQL database                           | `sudo mysql-create mydatabase`
| mysql-drop          | Drop MySQL database and user                        | `sudo mysql-drop`
| mysql-list          | List MySQL databases and users                      | `sudo mysql-list`
| ssl-create          | Generate Let's Encrypt SSL certificate              | `sudo ssl-create example.com`
| ssl-revoke          | Revoke Let's Encrypt SSL certificate                | `sudo ssl-revoke example.com`
| ssl-wildcard        | Generate wildcard Let's Encrypt SSL certificate     | `sudo ssl-wildcard example.com`

## TODO List

- [x] Change license from MIT to Apache 2.0
- [x] Write the change logs
- [x] Make each setup script as independent installer
- [ ] Add automatic installation wizard
- [ ] Fix pgAdmin4 installation
- [ ] Fix vhost proxy snippet
- [ ] Add more snippets

## License

Copyright (c) 2018-2019 Aris Ripandi

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at: <http://www.apache.org/licenses/LICENSE-2.0>

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.


changelog.md
## Changelog

All notable changes to this project will be documented here, the changelog
format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).

### [Unreleased]

### [4.0] - 2020/04/28
- Support Ubuntu Focal Fossa (20.04)

### [3.4] - 2019/11/14
- Changed project structure
- Enabling multiple distributiion support

### [3.3] - 2019/06/07
- Fix FTP server setup and configuration
- FTP server are optional (not installed by default)
- Change FTP server from ProFTPd to PureFTPd
- Fix Nginx configuration
- Add backup snippets
- Some other fixes

### [3.2] - 2019/05/20
- Add installation wizard
- Add FTP server setup
- Change Nginx repo to mainline

### [3.1] - 2019/04/09
- Added FTP server setup with SQL backend
- Added Mail server setup (TODO: fix it)
- Added snippet to create html vhost

### [3.0] - 2019/03/26
- Each setup script as independent installer
- Some improvements and fixes

### [2.4] - 2019/03/24
- Changed license from MIT to Apache 2.0
- Fix some configuration

### [2.3] - 2019/01/29
- Added Adminer as optional DB GUI
- Move phpMyAdmin to an alias
- Some configuration fixes

### [2.2] - 2019/01/13
- Add default user administartor
- Add feature for creating Ghost blogging platform
- Fix setup develoment environment
- Fix some nginx configuration
- Small fixes for some snippets
- Fix character encoding

### [2.1] - 2019/01/12
- Add development channel
- Fix default web page
- Fix default php page
- Fix some nginx configuration
- Some minor fixes and typo

### [2.0] - 2019/01/12
- Switch to Ubuntu LTS
- Nginx installer
- PHP FPM installer
- Python installer
- PostgreSQL installer
- MariaDB installer
- MySQL installer
- SSL Cert snippet
- MySQL snippet
- VirtualHost snippet
- Select databse version to install

### [1.0] - 2018/12/30
- Initial Release
- Nginx installer
- PHP FPM installer
- Python installer
- MariaDB installer
- VirtualHost snippet
- SSL Cert snippet
- MySQL snippet


wizard.sh
#!/bin/bash -e
if [[ $EUID -ne 0 ]]; then echo 'This script must be run as root' ; exit 1 ; fi

[ -z $ROOT ] && PWD=$(dirname "$(readlink -f "$0")") || PWD=$ROOT

# Functions
#-----------------------------------------------------------------------------------------

SetConfig() {
    crudini --set $PWD/config.ini $1 $2 $3
}

# Print welcome message
#-----------------------------------------------------------------------------------------
figlet "Are you ready?"
echo
while true; do
    read -ep "Type your answer ....... yes/no : " answer
    if [[ "${answer,,}" =~ ^(yes|y)$ ]]; then
        break
    elif [[ "${answer,,}" =~ ^(no|n)$ ]]; then
        exit 1
    fi
done
echo

# Basic questions
#-----------------------------------------------------------------------------------------
CFVAL=`crudini --get "$PWD/config.ini" system ssh_port`
read -ep "Please specify SSH port                        : " -i "$CFVAL" ssh_port
SetConfig system ssh_port $ssh_port

CFVAL=`crudini --get $PWD/config.ini system timezone`
read -ep "Please specify time zone                       : " -i "$CFVAL" timezone
SetConfig system timezone $timezone

CFVAL=`crudini --get $PWD/config.ini system disable_ipv6`
read -ep "Do you want to disable IPv6?            yes/no : " -i "$CFVAL" disable_ipv6
SetConfig system disable_ipv6 $disable_ipv6

# Determine total memory for swap usage
#-----------------------------------------------------------------------------------------
memoryTotal=`grep MemTotal /proc/meminfo | awk '{print $2}'`
if (( $memoryTotal >= 2097152 )); then opsi="no"; else opsi="yes";fi

CFVAL=`crudini --get $PWD/config.ini system swap_enable`
read -ep "Do you want to use Swap ?               yes/no : " -i "$opsi" swap_enable
SetConfig system swap_enable $swap_enable

if [[ "${swap_enable,,}" =~ ^(yes|y)$ ]] ; then
    CFVAL=`crudini --get $PWD/config.ini system swap_size`
    read -ep "Enter size of Swap (in megabyte)               : "  -i "$CFVAL" swap_size
    SetConfig system swap_size $swap_size
fi

# Telegram notification
#-----------------------------------------------------------------------------------------
CFVAL=`crudini --get $PWD/config.ini telegram enable`
read -ep "Use Telegram Notif                      yes/no : " -i "$CFVAL" sshnotify
SetConfig telegram enable $sshnotify

if [[ "${sshnotify,,}" =~ ^(yes|y)$ ]] ; then
    BOTKEY=`crudini --get $PWD/config.ini telegram bot_key`
    CHATID=`crudini --get $PWD/config.ini telegram chat_id`
    read -ep "Telegram Bot Key                               : " -i "$BOTKEY" tg_bot_key
    read -ep "Telegram User Chat ID                          : " -i "$CHATID" tg_chat_id
    SetConfig telegram bot_key $tg_bot_key
    SetConfig telegram chat_id $tg_chat_id
fi

# Nginx + Amplify
#-----------------------------------------------------------------------------------------
CFVAL=`crudini --get $PWD/config.ini nginx amplify`
read -ep "Install Nginx Amplify ?                 yes/no : " -i "$CFVAL" amplify_install
SetConfig nginx amplify $amplify_install

if [[ "${amplify_install,,}" =~ ^(yes|y)$ ]] ; then
    CFVAL=`crudini --get $PWD/config.ini nginx api_key`
    read -ep "Nginx Amplify API Key                          : " -i "$CFVAL" amplify_api
    SetConfig nginx api_key $amplify_api
fi

# Interpreter packages
#-----------------------------------------------------------------------------------------
CFVAL=`crudini --get $PWD/config.ini nodejs install`
read -ep "Install NodeJS ?                        yes/no : " -i "$CFVAL" nodejs_install
SetConfig nodejs install $nodejs_install

if [[ "${nodejs_install,,}" =~ ^(yes|y)$ ]] ; then
    CFVAL=`crudini --get $PWD/config.ini nodejs yarn`
    read -ep "Install Yarn ?                          yes/no : " -i "$CFVAL" yarn_install
    SetConfig nodejs yarn $yarn_install
fi

CFVAL=`crudini --get $PWD/config.ini php php56`
read -ep "Install PHP 5.6 ?                       yes/no : " -i "$CFVAL" php56_install
SetConfig php php56 $php56_install

CFVAL=`crudini --get $PWD/config.ini php php72`
read -ep "Install PHP 7.2 ?                       yes/no : " -i "$CFVAL" php72_install
SetConfig php php72 $php72_install

CFVAL=`crudini --get $PWD/config.ini php php73`
read -ep "Install PHP 7.3 ?                       yes/no : " -i "$CFVAL" php73_install
SetConfig php php73 $php73_install

CFVAL=`crudini --get $PWD/config.ini php default`
read -ep "Default PHP version?             (5.6/7.2/7.3) : " -i "$CFVAL" php_default
SetConfig php default $php_default

CFVAL=`crudini --get $PWD/config.ini python default`
read -ep "Default Python version?            (2.7 / 3.5) : " -i "$CFVAL" python_default
SetConfig python default $python_default

# MySQL / MariDB server
#-----------------------------------------------------------------------------------------
CFVAL=`crudini --get $PWD/config.ini mysql install`
read -ep "Install MySQL or MariaDB server ?       yes/no : " -i "$CFVAL" mysql_install
SetConfig mysql install $mysql_install

if [[ "${mysql_install,,}" =~ ^(yes|y)$ ]] ; then

    CFVAL=`crudini --get $PWD/config.ini mysql bind_address`
    read -ep "Database Bind Address                          : " -i "$CFVAL" bind_address
    SetConfig mysql bind_address $bind_address

    CFVAL=`crudini --get $PWD/config.ini mysql root_user`
    read -ep "Default MySQL database root user               : " -i "$CFVAL" mysql_user
    SetConfig mysql root_user $mysql_user

    CFVAL=`crudini --get $PWD/config.ini mysql root_pass`
    read -ep "Set database root password                     : "  -i "$CFVAL" root_pass
    if [[ "$root_pass" == "auto" ]] ; then
        SetConfig mysql root_pass $(openssl rand -base64 12 | tr -d "=+/" | cut -c1-25)
    else
        SetConfig mysql root_pass $root_pass
    fi

    CFVAL=`crudini --get $PWD/config.ini mysql engine`
    read -ep "Select database Engine         (mariadb/mysql) : " -i "$CFVAL" mysql_engine
    SetConfig mysql engine $mysql_engine

    if [[ "$mysql_engine" == "mysql" ]] ; then
        CFVAL=`crudini --get $PWD/config.ini mysql version`
        read -ep "Select MySQL version               (5.7 / 8.0) : " -i "$CFVAL" mysql_version
        SetConfig mysql version $mysql_version
    fi

fi

# PostgreSQL server
#-----------------------------------------------------------------------------------------
CFVAL=`crudini --get $PWD/config.ini postgres install`
read -ep "Install PostgreSQL database server ?    yes/no : " -i "$CFVAL" pgsql_install
SetConfig postgres install $pgsql_install

if [[ "${pgsql_install,,}" =~ ^(yes|y)$ ]] ; then

    CFVAL=`crudini --get $PWD/config.ini postgres bind_address`
    read -ep "Database Bind Address                          : " -i "$CFVAL" bind_address
    SetConfig postgres bind_address $bind_address

    # CFVAL=`crudini --get $PWD/config.ini postgres root_user`
    # read -ep "Default PostgreSQL root user                   : " -i "$CFVAL" pgsql_user
    # SetConfig postgres root_user $pgsql_user

    CFVAL=`crudini --get $PWD/config.ini postgres root_pass`
    read -ep "Set database root password                     : "  -i "$CFVAL" root_pass
    if [[ "$root_pass" == "auto" ]] ; then
        SetConfig postgres root_pass $(openssl rand -base64 12 | tr -d "=+/" | cut -c1-25)
    else
        SetConfig postgres root_pass $root_pass
    fi

    CFVAL=`crudini --get $PWD/config.ini postgres version`
    read -ep "Select PostgreSQL version      (9.6 / 10 / 11) : " -i "$CFVAL" pgsql_version
    SetConfig postgres version $pgsql_version

fi

# Redis server
#-----------------------------------------------------------------------------------------
CFVAL=`crudini --get $PWD/config.ini redis install`
read -ep "Install Redis Server ?                  yes/no : " -i "$CFVAL" redis_install
SetConfig redis install $redis_install

# Reboot after installation finish
#-----------------------------------------------------------------------------------------
CFVAL=`crudini --get $PWD/config.ini system reboot`
read -ep "Reboot server after install             yes/no : " -i "$CFVAL" reboot_after
SetConfig system reboot $reboot_after

# Mark wizard as finish
#-----------------------------------------------------------------------------------------
crudini --set $PWD/config.ini 'setup' 'ready' 'yes'
